lib/Parse/Lexer.cpp

```diff
@@ -1884,12 +1884,23 @@ void Lexer::lexStringLiteral(unsigned CustomDelimiterLen) {
         // Successfully scanned the body of the expression literal.
         ++CurPtr;
         continue;
-      } else if ((*CurPtr == '\r' || *CurPtr == '\n') && IsMultilineString) {
-        // The only case we reach here is unterminated single line string in the
-        // interpolation. For better recovery, go on after emitting an error.
-        diagnose(CurPtr, diag::lex_unterminated_string);
-        wasErroneous = true;
-        continue;
+      } else if (*CurPtr == '\r' || *CurPtr == '\n') {
+        if (IsMultilineString) {
+          // The only case we reach here is unterminated single line string in the
+          // interpolation. For better recovery, go on after emitting an error.
+          diagnose(CurPtr, diag::string_interpolation_unclosed);
+          diagnose(CurPtr, diag::lex_unterminated_string);
+          wasErroneous = true;
+          continue;
+        } else {
+          const char *fixItLoc = CurPtr - 1;
+          if (*fixItLoc == QuoteChar) {
+            fixItLoc -= CustomDelimiterLen;
+          }
+          diagnose(fixItLoc, diag::string_interpolation_unclosed);
```

Comment:
Why are we keeping the fix-it?

Reply:
Well why not? I don’t see there a reason for it not to exist in this context

Reply:
Think of it like the “insert ;” fix it for when 2 expressions are on the same line

Reply:
Have you seen the example in [this](https://github.com/apple/swift/pull/58882#discussion_r873227235) comment? In general, there is no way to predict the intended insertion location.

Reply:
Good point, will remove it then

Reply:
@AnthonyLatsis but wait, since fixItLoc is CurPtr subtracted by 1 and the CustomDelimiterLen, wouldn’t that example not apply?

Reply:
No, because `CurPtr` would point to the end of the line, not the intended end of the interpolation. The lexer is far from that smart, `skipToEndOfInterpolatedExpression` basically just skips until it hits a contextually invalid char or a closing paren.

Reply:
I see, should I keep the fixItLoc variable and rename it to interpolationLoc and remove the fix-it then?

Reply:
While we're here, are you up to adding the complementing `to match this opening '('` note?

Reply:
Oh yes! That was still in mind, I just wanted to make sure the changes I’ve made now already work

Reply:
> I see, should I keep the fixItLoc variable and rename it to interpolationLoc and remove the fix-it then?

Let's use `CurPtr`, that is where we realized the issue.

Reply:
But I’m confused, CurPtr would point outside the string, see: 
![](https://user-images.githubusercontent.com/48022799/168445857-bcd900d5-c49d-4ab2-9616-9f766788e548.png)

Reply:
Yeah, unclosed braces are handled similarly, albeit the error often points to the end of the file. I guess there is no simple way of devising whether we're inside the string, as in `let s = "\(x"; print(x)`, where `CurPtr` points the end of the line.

Reply:
Well I guess that’s fine now that the fix it is gone, I’ll change it to CurPtr then